File: evaluation/go/segment_conformance_test.go
// Copyright 2026 The Bucketeer Authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package evaluation

import (
	"encoding/json"
	"os"
	"path/filepath"
	"testing"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"

	ftproto "github.com/bucketeer-io/bucketeer/v2/proto/feature"
	userproto "github.com/bucketeer-io/bucketeer/v2/proto/user"
)

// The conformance fixtures are shared with evaluation/typescript so the two
// engines stay in lockstep. See evaluation/testdata/segment_rules_conformance.json.
const conformanceFixturePath = "../testdata/segment_rules_conformance.json"

type conformanceClause struct {
	ID        string   `json:"id"`
	Attribute string   `json:"attribute"`
	Operator  string   `json:"operator"`
	Values    []string `json:"values"`
}

type conformanceRule struct {
	ID      string              `json:"id"`
	Clauses []conformanceClause `json:"clauses"`
}

type conformanceSegment struct {
	ID    string            `json:"id"`
	Rules []conformanceRule `json:"rules"`
}

type conformanceSegmentUser struct {
	SegmentID string `json:"segmentId"`
	UserID    string `json:"userId"`
	State     string `json:"state"`
}

type conformanceUser struct {
	ID   string            `json:"id"`
	Data map[string]string `json:"data"`
}

type conformanceTestCase struct {
	Desc       string          `json:"desc"`
	SegmentIDs []string        `json:"segmentIds"`
	User       conformanceUser `json:"user"`
	Expected   bool            `json:"expected"`
}

type conformanceFixture struct {
	Segments     []conformanceSegment     `json:"segments"`
	SegmentUsers []conformanceSegmentUser `json:"segmentUsers"`
	TestCases    []conformanceTestCase    `json:"testCases"`
}

func TestSegmentRulesConformance(t *testing.T) {
	t.Parallel()
	fixture := loadConformanceFixture(t)
	segments := make(map[string]*ftproto.Segment, len(fixture.Segments))
	for _, s := range fixture.Segments {
		segments[s.ID] = &ftproto.Segment{
			Id:    s.ID,
			Rules: toProtoRules(t, s.Rules),
		}
	}
	segmentUsers := make([]*ftproto.SegmentUser, 0, len(fixture.SegmentUsers))
	for _, su := range fixture.SegmentUsers {
		state, ok := ftproto.SegmentUser_State_value[su.State]
		require.True(t, ok, "unknown segment user state: %s", su.State)
		segmentUsers = append(segmentUsers, &ftproto.SegmentUser{
			SegmentId: su.SegmentID,
			UserId:    su.UserID,
			State:     ftproto.SegmentUser_State(state),
		})
	}
	evaluator := &segmentEvaluator{}
	for _, tc := range fixture.TestCases {
		tc := tc
		t.Run(tc.Desc, func(t *testing.T) {
			t.Parallel()
			user := &userproto.User{Id: tc.User.ID, Data: tc.User.Data}
			actual, err := evaluator.Evaluate(tc.SegmentIDs, user, segments, segmentUsers)
			assert.NoError(t, err)
			assert.Equal(t, tc.Expected, actual)
		})
	}
}

func loadConformanceFixture(t *testing.T) *conformanceFixture {
	t.Helper()
	data, err := os.ReadFile(filepath.Clean(conformanceFixturePath))
	require.NoError(t, err)
	fixture := &conformanceFixture{}
	require.NoError(t, json.Unmarshal(data, fixture))
	return fixture
}

func toProtoRules(t *testing.T, rules []conformanceRule) []*ftproto.Rule {
	t.Helper()
	protoRules := make([]*ftproto.Rule, 0, len(rules))
	for _, r := range rules {
		clauses := make([]*ftproto.Clause, 0, len(r.Clauses))
		for _, c := range r.Clauses {
			operator, ok := ftproto.Clause_Operator_value[c.Operator]
			require.True(t, ok, "unknown clause operator: %s", c.Operator)
			clauses = append(clauses, &ftproto.Clause{
				Id:        c.ID,
				Attribute: c.Attribute,
				Operator:  ftproto.Clause_Operator(operator),
				Values:    c.Values,
			})
		}
		protoRules = append(protoRules, &ftproto.Rule{
			Id:      r.ID,
			Clauses: clauses,
		})
	}
	return protoRules
}

File: evaluation/typescript/src/__tests__/segment_conformance_test.ts
import test from 'ava';
import * as fs from 'fs';
import * as path from 'path';
import { Segment, SegmentUser } from '../proto/feature/segment_pb';
import { Rule } from '../proto/feature/rule_pb';
import { Clause } from '../proto/feature/clause_pb';
import { SegmentEvaluator } from '../segmentEvaluator';
import { createUser } from '../modelFactory';

// The conformance fixtures are shared with evaluation/go so the two engines
// stay in lockstep. See evaluation/testdata/segment_rules_conformance.json.
interface ConformanceClause {
  id: string;
  attribute: string;
  operator: string;
  values: string[];
}

interface ConformanceRule {
  id: string;
  clauses: ConformanceClause[];
}

interface ConformanceSegment {
  id: string;
  rules: ConformanceRule[];
}

interface ConformanceSegmentUser {
  segmentId: string;
  userId: string;
  state: string;
}

interface ConformanceTestCase {
  desc: string;
  segmentIds: string[];
  user: { id: string; data: { [key: string]: string } };
  expected: boolean;
}

interface ConformanceFixture {
  segments: ConformanceSegment[];
  segmentUsers: ConformanceSegmentUser[];
  testCases: ConformanceTestCase[];
}

function loadFixture(): ConformanceFixture {
  // Compiled tests run from __test/__tests__, source runs from src/__tests__.
  const candidates = [
    path.join(__dirname, '../../../testdata/segment_rules_conformance.json'),
    path.join(__dirname, '../../testdata/segment_rules_conformance.json'),
  ];
  for (const candidate of candidates) {
    if (fs.existsSync(candidate)) {
      return JSON.parse(fs.readFileSync(candidate, 'utf-8'));
    }
  }
  throw new Error('segment_rules_conformance.json not found');
}

function toOperator(name: string): Clause.OperatorMap[keyof Clause.OperatorMap] {
  const operator = Clause.Operator[name as keyof Clause.OperatorMap];
  if (operator === undefined) {
    throw new Error(`unknown clause operator: ${name}`);
  }
  return operator;
}

function toState(name: string): SegmentUser.StateMap[keyof SegmentUser.StateMap] {
  const state = SegmentUser.State[name as keyof SegmentUser.StateMap];
  if (state === undefined) {
    throw new Error(`unknown segment user state: ${name}`);
  }
  return state;
}

function toProtoRules(rules: ConformanceRule[]): Rule[] {
  return rules.map((r) => {
    const rule = new Rule();
    rule.setId(r.id);
    rule.setClausesList(
      r.clauses.map((c) => {
        const clause = new Clause();
        clause.setId(c.id);
        clause.setAttribute(c.attribute);
        clause.setOperator(toOperator(c.operator));
        clause.setValuesList(c.values);
        return clause;
      }),
    );
    return rule;
  });
}

const fixture = loadFixture();

const segments = new Map<string, Segment>();
fixture.segments.forEach((s) => {
  const segment = new Segment();
  segment.setId(s.id);
  segment.setRulesList(toProtoRules(s.rules));
  segments.set(s.id, segment);
});

const segmentUsers = fixture.segmentUsers.map((su) => {
  const segmentUser = new SegmentUser();
  segmentUser.setSegmentId(su.segmentId);
  segmentUser.setUserId(su.userId);
  segmentUser.setState(toState(su.state));
  return segmentUser;
});

fixture.testCases.forEach((tc) => {
  test(`conformance: ${tc.desc}`, (t) => {
    const evaluator = new SegmentEvaluator();
    const user = createUser(tc.user.id, tc.user.data);
    const actual = evaluator.evaluate(tc.segmentIds, user, segments, segmentUsers);
    t.is(actual, tc.expected);
  });
});
